lambdas/access_counts/index.py

```diff
@@ -190,20 +267,65 @@ def delete_temp_dir():
+    end_ts = datetime.utcnow() - timedelta(minutes=5)
+
+    # Start of the CloudTrail time range: the end timestamp from the previous run, or a year ago if it's the first run.
+    try:
+        start_ts = datetime.utcfromtimestamp(float(s3.get_object(Bucket=QUERY_RESULT_BUCKET, Key=LAST_UPDATE_KEY)['Body'].read()))
+    except s3.exceptions.NoSuchKey as ex:
+        start_ts = end_ts - timedelta(days=365)
+
+    start_date = start_ts.date()
+    end_date = end_ts.date()
+
+    # Delete the temporary directory where Athena query results are written to.
+    delete_dir(QUERY_RESULT_BUCKET, QUERY_TEMP_DIR)
+
+    # Create a CloudTrail table, but only with partitions for the last N days, to avoid scanning all of the data.
+    # A bucket can have data for multiple accounts and multiple regions, so those need to be handled first.
+    partition_queries = []
+    for account_response in s3.list_objects_v2(Bucket=CLOUDTRAIL_BUCKET, Prefix='AWSLogs/', Delimiter='/').get('CommonPrefixes') or []:
+        account = account_response['Prefix'].split('/')[1]
+        for region_response in s3.list_objects_v2(Bucket=CLOUDTRAIL_BUCKET, Prefix=f'AWSLogs/{account}/CloudTrail/', Delimiter='/').get('CommonPrefixes') or []:
+            region = region_response['Prefix'].split('/')[3]
+            date = start_date
+            while date <= end_date:
+                query = ADD_CLOUDTRAIL_PARTITION.format(
+                    account=account,
+                    region=region,
+                    year=date.year,
+                    month=date.month,
+                    day=date.day
+                )
+                partition_queries.append(query)
+                date += timedelta(days=1)
+
+    # Drop old Athena tables from previous runs.
     # (They're in the DB owned by the stack, so safe to do.)
-    for query_id in [run_query(DROP_OBJECT_ACCESS_LOG), run_query(DROP_PACKAGE_HASHES)]:
+    for query_id in [run_query(DROP_CLOUDTRAIL), run_query(DROP_OBJECT_ACCESS_LOG), run_query(DROP_PACKAGE_HASHES)]:
         wait_for_query(query_id)
 
-    for query_id in [run_query(CREATE_OBJECT_ACCESS_LOG), run_query(CREATE_PACKAGE_HASHES)]:
+    # Create new Athena tables.
+    for query_id in [run_query(CREATE_CLOUDTRAIL), run_query(CREATE_OBJECT_ACCESS_LOG), run_query(CREATE_PACKAGE_HASHES)]:
         wait_for_query(query_id)
 
+    # Create CloudTrail partitions, after the CloudTrail table is created.
+    for query_id in [run_query(q) for q in partition_queries]:
+        wait_for_query(query_id)
+
+    # Scan CloudTrail and insert new data into "object_access_log".
+    insert_query = INSERT_INTO_OBJECT_ACCESS_LOG.format(start_ts=start_ts.timestamp(), end_ts=end_ts.timestamp())
+    wait_for_query(run_query(insert_query))
+
+    # Save the end timestamp.
+    s3.put_object(Bucket=QUERY_RESULT_BUCKET, Key=LAST_UPDATE_KEY, Body=str(end_ts.timestamp()), ContentType='text/plain')
```

Comment:
Just to be clear: we don't have object versioning on in `QUERY_RESULT_BUCKET`? And we don't expect to hit any consistency issues?

Reply:
No versioning. Shouldn't hit any issues as long as there are no overlapping lambdas. It runs once an hour, so should be safe?